go/saltpack/verify.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package saltpack
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+)
+
+// NewVerifyStream creates a stream that consumes data from reader
+// r.  It returns the signer's public key and a reader that only
+// contains verified data.  If the signer's key is not in keyring,
+// it will return an error.
+func NewVerifyStream(r io.Reader, keyring SigKeyring) (skey SigningPublicKey, vs io.Reader, err error) {
+	s := newVerifyStream(r)
+	hdr, err := s.readHeader(MessageTypeAttachedSignature)
+	if err != nil {
+		return nil, nil, err
+	}
+	skey = keyring.LookupSigningPublicKey(hdr.SenderPublic)
+	if skey == nil {
+		return nil, nil, ErrNoSenderKey
+	}
+	s.publicKey = skey
+	return skey, s, nil
+}
+
+// Verify checks the signature in signedMsg.  It returns the
+// signer's public key and a verified message.
+func Verify(signedMsg []byte, keyring SigKeyring) (skey SigningPublicKey, verifiedMsg []byte, err error) {
+	skey, stream, err := NewVerifyStream(bytes.NewReader(signedMsg), keyring)
+	if err != nil {
+		return nil, nil, err
+	}
+	if stream == nil {
+		return nil, nil, ErrNoStream
+	}
+	if skey == nil {
+		return nil, nil, ErrNoSenderKey
+	}
+
+	verifiedMsg, err = ioutil.ReadAll(stream)
+	if err != nil {
+		return nil, nil, err
+	}
+	return skey, verifiedMsg, nil
+}
+
+// VerifyDetached verifies that signature is a valid signature for
+// message, and that the public key for the signer is in keyring.
+// It returns the signer's public key.
+func VerifyDetached(message, signature []byte, keyring SigKeyring) (skey SigningPublicKey, err error) {
+	s := newVerifyStream(bytes.NewBuffer(signature))
+	hdr, err := s.readHeader(MessageTypeDetachedSignature)
+	if err != nil {
+		return nil, err
+	}
+	if len(hdr.Signature) == 0 {
+		return nil, ErrNoDetachedSignature
+	}
+	skey = keyring.LookupSigningPublicKey(hdr.SenderPublic)
+	if skey == nil {
+		return nil, ErrNoSenderKey
+	}
+
+	ok, err := skey.Verify(computeDetachedDigest(hdr.Nonce, message), hdr.Signature)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, ErrBadSignature
+	}
+
+	return skey, nil
+}
+
+type verifyStream struct {
+	stream    *framedMsgpackStream
+	state     readState
+	buffer    []byte
+	header    *SignatureHeader
+	publicKey SigningPublicKey
+}
+
+func newVerifyStream(r io.Reader) *verifyStream {
+	return &verifyStream{
+		stream: newFramedMsgpackStream(r),
```

Comment:
Can you rename this to `newMsgpackStream`.  I forgot to rename earlier, since we got rid of framing for saltpack.